### ISSUE TITLE
UnitTests fails with segmentation fault on Odroid-XU4

### ISSUE THREAD

**Issue by vn218** [ARM-software/armnn#696]
I built ARM-NN library for my odroid xu4 (aarch32) by following the ARM NN Build Tool tutorial. 
On running UnitTests, it fails with segmentation error.

```
odroid@odroid:~/Desktop/aarch32_build$ export LD_LIBRARY_PATH=.;./UnitTests 
[doctest] doctest version is "2.4.6"
[doctest] run with "--help" for options
===============================================================================
/home/arm-user/source/armnn/src/armnn/test/optimizations/FuseActivationTests.cpp:878:
TEST SUITE: Optimizer
TEST CASE:  FuseReLUIntoConvFloat32GpuAccTest

/home/arm-user/source/armnn/src/armnn/test/optimizations/FuseActivationTests.cpp:878: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal

===============================================================================
[doctest] test cases:    403 |    402 passed | 1 failed | 2284 skipped
[doctest] assertions: 772095 | 772095 passed | 0 failed |
[doctest] Status: FAILURE!
Segmentation fault
```
I tried increasing stack size but it still fails. I am running Ubuntu Mate 22.04

**Comment by james-conroy-arm** [comment]
Hi @vn218 ,

Many thanks for raising this issue. Did you build Arm NN using with build-tool with or without Docker? On a remote machine or on the Odroid board itself?

Since there is a segfault, there may be other tests failing that are not executing. To help us try narrow down the issue, could you please provide us with the output for the following UnitTests flags:

Run FuseReLUIntoConvFloat32GpuAccTest exclusively:
export LD_LIBRARY_PATH=.;./UnitTests --test-case=FuseReLUIntoConvFloat32GpuAccTest

Run all tests except FuseReLUIntoConvFloat32GpuAccTest:
export LD_LIBRARY_PATH=.;./UnitTests --test-case-exclude=FuseReLUIntoConvFloat32GpuAccTest

Run all test suites except Optimizer tests:
export LD_LIBRARY_PATH=.;./UnitTests --test-suite-exclude=Optimizer

Run all tests except those on the GpuAcc backend:
export LD_LIBRARY_PATH=.;./UnitTests --test-case-exclude=*GpuAcc*

**Comment by james-conroy-arm** [comment]
Thanks for that @vn218 . Looks like many tests are failing on GPU.

Could you please try this more specific GPU exclusion filter:
export LD_LIBRARY_PATH=.;./UnitTests --test-case-exclude=\*Gpu\*

You could also try adding Mali to your LD_LIBRARY_PATH, but this may not be the exact location on your device (this works on an aarch64 Odroid N2+ board):
export LD_LIBRARY_PATH=.:/usr/share/mali; ./UnitTests

James

**Comment by vn218** [comment]
The output ....
```
odroid@odroid:~/Desktop/aarch32_build$ export LD_LIBRARY_PATH=.;./UnitTests --test-case-exclude=*GPU*
[doctest] doctest version is "2.4.6"
[doctest] run with "--help" for options
Warning: Timed out waiting on profiling service activation for 3000.12 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
===============================================================================
/home/arm-user/source/armnn/src/backends/cl/test/ClCreateWorkloadTests.cpp:226:
TEST SUITE: CreateWorkloadCl
TEST CASE:  CreateBatchNormalizationFloatNchwWorkload

/home/arm-user/source/armnn/src/backends/cl/test/ClCreateWorkloadTests.cpp:226: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal

===============================================================================
[doctest] test cases:    537 |    536 passed | 1 failed | 2150 skipped
[doctest] assertions: 773224 | 773224 passed | 0 failed |
[doctest] Status: FAILURE!
Segmentation fault
```
Adding mali to LD_LIBRARY_PATH didn't help

**Comment by james-conroy-arm** [comment]
Please try:
`export LD_LIBRARY_PATH=.;./UnitTests --test-case-exclude=*GPU*,*CL*`

What is the contents of `/usr/share/mali/` ?

**Comment by vn218** [comment]
adding --test-case-exclude=*Cl* gives the same output as before, however adding --test-suite-exclude=*Cl* passes
```
odroid@odroid:~/Desktop/aarch32_build$ export LD_LIBRARY_PATH=.;./UnitTests --test-case-exclude=*GPU*,*Cl* --test-suite-exclude=*Cl*
[doctest] doctest version is "2.4.6"
[doctest] run with "--help" for options
Warning: Timed out waiting on profiling service activation for 3000.11 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
Warning: Timed out waiting on profiling service activation for 3000.11 ms
===============================================================================
[doctest] test cases:   1592 |   1592 passed | 0 failed | 1095 skipped
[doctest] assertions: 780221 | 780221 passed | 0 failed |
[doctest] Status: SUCCESS!
```
contents of mali .....
```
odroid@odroid:~/Desktop/aarch32_build$ ls -R /usr/share/mali/
/usr/share/mali/:
headers  pkgconfig

/usr/share/mali/headers:
CL  CL_1_2  CL_2_0  EGL  GLES  GLES2  GLES3  KHR

/usr/share/mali/headers/CL:
cl.h        cl_egl.h  cl_gl.h      cl_platform.h
cl_d3d10.h  cl_ext.h  cl_gl_ext.h  opencl.h

/usr/share/mali/headers/CL_1_2:
CL

/usr/share/mali/headers/CL_1_2/CL:
cl.h        cl_d3d11.h              cl_egl.h  cl_gl.h      cl_platform.h
cl_d3d10.h  cl_dx9_media_sharing.h  cl_ext.h  cl_gl_ext.h  opencl.h

/usr/share/mali/headers/CL_2_0:
CL

/usr/share/mali/headers/CL_2_0/CL:
cl.h    cl_d3d10.h  cl_dx9_media_sharing.h  cl_ext.h  cl_gl_ext.h    opencl.h
cl.hpp  cl_d3d11.h  cl_egl.h                cl_gl.h   cl_platform.h

/usr/share/mali/headers/EGL:
egl.h  eglext.h  eglplatform.h

/usr/share/mali/headers/GLES:
egl.h  gl.h  glext.h  glplatform.h

/usr/share/mali/headers/GLES2:
gl2.h  gl2ext.h  gl2platform.h

/usr/share/mali/headers/GLES3:
gl3.h  gl31.h  gl32.h  gl3platform.h

/usr/share/mali/headers/KHR:
khrplatform.h

/usr/share/mali/pkgconfig:
egl.pc  glesv1.pc  glesv1_cm.pc  glesv2.pc  glesv3.pc  opencl.pc
```

**Comment by vn218** [comment]
Also, when I try to run a tflite model (yolov5s), it fails with segfault for all combinations of backends.

**Comment by vn218** [comment]
@james-conroy-arm Hi! Is there any update?

**Comment by james-conroy-arm** [comment]
Hi @vn218 ,

The Mali driver on your device may not be setup correctly, it's difficult for me to say.

Albeit on a different Odroid board (N2+) and aarch64, this is the contents of /usr/share/mali:

**Comment by vn218** [comment]
@james-conroy-arm Thanks! Linking libMali.so and libOpenCL.so worked (libMali.so was at a different location). But, now I am getting a new set of errors .........
[UnitTests.log](https://github.com/ARM-software/armnn/files/9820422/UnitTests.log)

But all the DelegateUnitTests are passing, so I guess that is enough for my use case.

yolov5s is still not working, I've raised a separate issue for it.

**Comment by FrancisMurtagh-arm** [comment]
Closing as original issue resolved, and new issue raised.
